core/commands/profile.go

```diff
@@ -0,0 +1,234 @@
+- Any modifications you've made to go-ipfs.
+`,
+	},
+	NoLocal: true,
+	Options: []cmds.Option{
+		cmds.StringOption(outputOptionName, "o", "The path where the output should be stored."),
+		cmds.StringOption(cpuProfileTimeOption, "The amount of time spent profiling CPU usage.").WithDefault("30s"),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		cpuProfileTimeStr, _ := req.Options[cpuProfileTimeOption].(string)
+		cpuProfileTime, err := time.ParseDuration(cpuProfileTimeStr)
+		if err != nil {
+			return fmt.Errorf("failed to parse CPU profile duration %q: %w", cpuProfileTimeStr, err)
+		}
+
+		r, w := io.Pipe()
+		go func() {
+			_ = w.CloseWithError(writeProfiles(req.Context, cpuProfileTime, w))
+		}()
+		return res.Emit(r)
+	},
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(res cmds.Response, re cmds.ResponseEmitter) error {
+			v, err := res.Next()
+			if err != nil {
+				return err
+			}
+
+			outReader, ok := v.(io.Reader)
+			if !ok {
+				return e.New(e.TypeErr(outReader, v))
+			}
+
+			outPath, _ := res.Request().Options[outputOptionName].(string)
+			if outPath == "" {
+				outPath = "ipfs-profile-" + time.Now().Format(timeFormat) + ".zip"
+			}
+			fi, err := os.Create(outPath)
+			if err != nil {
+				return err
+			}
+			defer fi.Close()
+
+			_, err = io.Copy(fi, outReader)
+			if err != nil {
+				return err
+			}
+			return re.Emit(&profileResult{File: outPath})
+		},
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *profileResult) error {
+			fmt.Fprintf(w, "Wrote profiles to: %s\n", out.File)
+			return nil
+		}),
+	},
+}
+
+func writeProfiles(ctx context.Context, cpuProfileTime time.Duration, w io.Writer) error {
+	archive := zip.NewWriter(w)
+
+	// Take some profiles.
+	type profile struct {
+		name  string
+		file  string
+		debug int
+	}
+
+	profiles := []profile{{
+		name:  "goroutine",
+		file:  "goroutines.stacks",
+		debug: 2,
+	}, {
+		name: "goroutine",
+		file: "goroutines.pprof",
+	}, {
+		name: "heap",
+		file: "heap.pprof",
+	}}
+
+	for _, profile := range profiles {
+		prof := pprof.Lookup(profile.name)
+		out, err := archive.Create(profile.file)
+		if err != nil {
+			return err
+		}
+		err = prof.WriteTo(out, profile.debug)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Take a CPU profile.
+	if cpuProfileTime != 0 {
+		out, err := archive.Create("cpu.pprof")
+		if err != nil {
+			return err
+		}
+
+		err = writeCPUProfile(ctx, cpuProfileTime, out)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Collect version info
+	// I'd use diag sysinfo, but that includes some more sensitive information
+	// (GOPATH, etc.).
+	{
+		out, err := archive.Create("version.json")
+		if err != nil {
+			return err
+		}
+
+		err = json.NewEncoder(out).Encode(getVersionInfo())
+		if err != nil {
+			return err
+		}
+	}
+
+	// Collect binary
+	if fi, err := openIPFSBinary(); err == nil {
```

Comment:
I don't know if this is worth doing by default. We have the version info and it bloats the size of the zip. WDYT?

Reply:
We need it for the symbols in case the user deployed a custom binary. I guess we could include it iff the version is "dirty"?

Note: the zip is ~26MiB. So it's not terrible.

Reply:
It's not, although I think I've seen maximum sizes for emails and Slack attachments from external parties be lower. I'm not sure what the maximum GitHub issue attachment size is.

Even if it's not the default we should probably have a flag here just in case